lib/internal/url.js

```diff
@@ -67,48 +67,151 @@ class TupleOrigin {
   }
 
   toString(unicode = false) {
-    var result = this.scheme;
+    var result = this[kScheme];
     result += '://';
-    result += unicode ? URL.domainToUnicode(this.host) : this.host;
-    if (this.port !== undefined && this.port !== null)
-      result += `:${this.port}`;
+    result += unicode ? URL.domainToUnicode(this[kHost]) : this[kHost];
+    if (this[kPort] !== undefined && this[kPort] !== null)
+      result += `:${this[kPort]}`;
     return result;
   }
 
   inspect() {
     return `TupleOrigin {
-      scheme: ${this.scheme},
-      host: ${this.host},
-      port: ${this.port},
-      domain: ${this.domain}
+      scheme: ${this[kScheme]},
+      host: ${this[kHost]},
+      port: ${this[kPort]},
+      domain: ${this[kDomain]}
     }`;
   }
 }
 
+function onParseComplete(flags, protocol, username, password,
+                         host, port, path, query, fragment) {
+  if (flags & binding.URL_FLAGS_FAILED)
+    throw new TypeError('Invalid URL');
+  const ctx = this[context];
+  ctx.flags = flags;
+  ctx.scheme = protocol;
+  ctx.username = username;
+  ctx.password = password;
+  ctx.port = port;
+  ctx.path = path;
+  ctx.query = query;
+  ctx.fragment = fragment;
+  ctx.host = host;
+  this[searchParams] = new URLSearchParams(query);
+  this[searchParams][context] = this;
+}
+
+function onParseProtocolComplete(flags, protocol, username, password,
+                                 host, port, path, query, fragment) {
+  if (flags & binding.URL_FLAGS_FAILED)
+    return;
+  const newIsSpecial = (flags & binding.URL_FLAGS_SPECIAL) != 0;
+  const s = this[special];
+  const ctx = this[context];
+  if ((s && !newIsSpecial) ||
+      (!s && newIsSpecial) ||
```

Comment:
this[special] is guaranteed to be a boolean. `s` and `!s` are sufficient. There's no need (and no benefit) for getting fancy with it.